File: clients/go/flrig_control.go
package main

import (
	"bytes"
	"encoding/xml"
	"fmt"
	"io"
	"log"
	"net/http"
	"strconv"
	"sync"
	"time"
)

// FlrigClient represents an XML-RPC client for flrig radio control
type FlrigClient struct {
	host      string
	port      int
	vfo       string // "A" or "B"
	client    *http.Client
	connected bool
	mu        sync.RWMutex

	// Cached values for quick access (updated by polling)
	cachedFrequency int
	cachedMode      string
	cachedPTT       bool
	cacheMu         sync.RWMutex
	firstPoll       bool // Track if this is the first poll after connection

	// Callbacks
	frequencyCallback func(int)
	modeCallback      func(string)
	pttCallback       func(bool)
	errorCallback     func(string)

	// Debug
	pollCount int
}

// NewFlrigClient creates a new flrig XML-RPC client
func NewFlrigClient(host string, port int, vfo string) *FlrigClient {
	return &FlrigClient{
		host: host,
		port: port,
		vfo:  vfo,
		client: &http.Client{
			Timeout: 5 * time.Second,
		},
		connected: false,
		firstPoll: true, // Mark that we haven't polled yet
	}
}

// Connect establishes connection to flrig server
func (f *FlrigClient) Connect() error {
	f.mu.Lock()
	defer f.mu.Unlock()

	// Test connection by listing methods
	_, err := f.callMethod("system.listMethods", []interface{}{})
	if err != nil {
		return fmt.Errorf("failed to connect to flrig at %s:%d: %w", f.host, f.port, err)
	}

	f.connected = true
	return nil
}

// Disconnect closes the connection
func (f *FlrigClient) Disconnect() {
	f.mu.Lock()
	defer f.mu.Unlock()
	f.connected = false
}

// IsConnected returns connection status
func (f *FlrigClient) IsConnected() bool {
	f.mu.RLock()
	defer f.mu.RUnlock()
	return f.connected
}

// GetFrequency returns the current frequency in Hz
func (f *FlrigClient) GetFrequency() (int, error) {
	f.mu.RLock()
	if !f.connected {
		f.mu.RUnlock()
		return 0, fmt.Errorf("not connected to flrig")
	}
	f.mu.RUnlock()

	result, err := f.callMethod("rig.get_vfo", []interface{}{})
	if err != nil {
		// Don't mark as disconnected - let connection errors be handled elsewhere
		return 0, err
	}

	// Handle different response types - flrig can return string or number
	var freq float64
	switch v := result.(type) {
	case string:
		freq, err = strconv.ParseFloat(v, 64)
		if err != nil {
			return 0, fmt.Errorf("failed to parse frequency string: %w", err)
		}
	case float64:
		freq = v
	case int:
		freq = float64(v)
	case int64:
		freq = float64(v)
	default:
		return 0, fmt.Errorf("unexpected response type for frequency: %T", result)
	}

	return int(freq), nil
}

// SetFrequency sets the frequency in Hz
func (f *FlrigClient) SetFrequency(freqHz int) error {
	f.mu.RLock()
	if !f.connected {
		f.mu.RUnlock()
		return fmt.Errorf("not connected to flrig")
	}
	f.mu.RUnlock()

	_, err := f.callMethod("rig.set_vfo", []interface{}{float64(freqHz)})
	if err != nil {
		// Don't mark as disconnected - let connection errors be handled elsewhere
		return err
	}

	return nil
}

// GetMode returns the current mode
func (f *FlrigClient) GetMode() (string, error) {
	f.mu.RLock()
	if !f.connected {
		f.mu.RUnlock()
		return "", fmt.Errorf("not connected to flrig")
	}
	f.mu.RUnlock()

	result, err := f.callMethod("rig.get_mode", []interface{}{})
	if err != nil {
		// Don't mark as disconnected - let connection errors be handled elsewhere
		return "", err
	}

	mode, ok := result.(string)
	if !ok {
		return "", fmt.Errorf("unexpected response type for mode")
	}

	if mode == "" {
		return "Unknown", nil
	}

	return mode, nil
}

// SetMode sets the mode (USB, LSB, CW, AM, FM, etc.)
func (f *FlrigClient) SetMode(mode string) error {
	f.mu.RLock()
	if !f.connected {
		f.mu.RUnlock()
		return fmt.Errorf("not connected to flrig")
	}
	f.mu.RUnlock()

	_, err := f.callMethod("rig.set_mode", []interface{}{mode})
	if err != nil {
		// Don't mark as disconnected - let connection errors be handled elsewhere
		return err
	}

	return nil
}

// GetPTT returns the PTT (Push-To-Talk) status
func (f *FlrigClient) GetPTT() (bool, error) {
	f.mu.RLock()
	if !f.connected {
		f.mu.RUnlock()
		return false, fmt.Errorf("not connected to flrig")
	}
	f.mu.RUnlock()

	result, err := f.callMethod("rig.get_ptt", []interface{}{})
	if err != nil {
		// Don't mark as disconnected - let connection errors be handled elsewhere
		return false, err
	}

	// Result can be bool or string from our parser
	switch v := result.(type) {
	case bool:
		return v, nil
	case string:
		// Parse string representation - can be "0"/"1" or "true"/"false"
		return v == "1" || v == "true", nil
	default:
		return false, fmt.Errorf("unexpected response type for PTT: %T", result)
	}
}

// SetPTT sets the PTT state
func (f *FlrigClient) SetPTT(state bool) error {
	f.mu.RLock()
	if !f.connected {
		f.mu.RUnlock()
		return fmt.Errorf("not connected to flrig")
	}
	f.mu.RUnlock()

	pttValue := 0
	if state {
		pttValue = 1
	}

	_, err := f.callMethod("rig.set_ptt", []interface{}{pttValue})
	if err != nil {
		// Don't mark as disconnected - let connection errors be handled elsewhere
		return err
	}

	return nil
}

// SetVFO switches to the specified VFO (A or B)
func (f *FlrigClient) SetVFO(vfo string) error {
	f.mu.RLock()
	if !f.connected {
		f.mu.RUnlock()
		return fmt.Errorf("not connected to flrig")
	}
	f.mu.RUnlock()

	_, err := f.callMethod("rig.set_AB", []interface{}{vfo})
	if err != nil {
		// Don't mark as disconnected - let connection errors be handled elsewhere
		return err
	}

	f.mu.Lock()
	f.vfo = vfo
	f.mu.Unlock()

	return nil
}

// GetVFO returns the current VFO
func (f *FlrigClient) GetVFO() string {
	f.mu.RLock()
	defer f.mu.RUnlock()
	return f.vfo
}

// Poll updates all cached values and triggers callbacks
func (f *FlrigClient) Poll() {
	// Get frequency
	freq, err := f.GetFrequency()
	if err != nil {
		log.Printf("flrig: Poll() GetFrequency error: %v", err)
		if f.errorCallback != nil {
			f.errorCallback(err.Error())
		}
		return
	}

	// Get mode
	mode, err := f.GetMode()
	if err != nil {
		log.Printf("flrig: Poll() GetMode error: %v", err)
		if f.errorCallback != nil {
			f.errorCallback(err.Error())
		}
		return
	}

	// Get PTT
	ptt, err := f.GetPTT()
	if err != nil {
		log.Printf("flrig: Poll() GetPTT error: %v", err)
		if f.errorCallback != nil {
			f.errorCallback(err.Error())
		}
		return
	}

	// Update cache and trigger callbacks
	f.cacheMu.Lock()
	oldFreq := f.cachedFrequency
	oldMode := f.cachedMode
	oldPTT := f.cachedPTT

	f.cachedFrequency = freq
	f.cachedMode = mode
	f.cachedPTT = ptt
	f.cacheMu.Unlock()

	// Debug: log current values every 10 seconds (20 polls at 500ms)
	f.pollCount++
	if f.pollCount%20 == 0 {
		log.Printf("flrig: Poll #%d - freq=%d, mode=%s, ptt=%v", f.pollCount, freq, mode, ptt)
	}

	// On first poll, always trigger callbacks to sync initial state
	// On subsequent polls, only trigger if values changed
	isFirstPoll := f.firstPoll
	if isFirstPoll {
		f.firstPoll = false
		log.Printf("flrig: First poll - initializing with freq=%d, mode=%s, ptt=%v", freq, mode, ptt)
	}

	// Trigger callbacks if values changed OR if this is the first poll
	if freq != oldFreq || isFirstPoll {
		if freq != oldFreq {
			log.Printf("flrig: frequency changed from %d to %d Hz", oldFreq, freq)
		}
		if f.frequencyCallback != nil {
			f.frequencyCallback(freq)
		} else {
			log.Printf("flrig: frequency callback is nil!")
		}
	}
	if mode != oldMode || isFirstPoll {
		if mode != oldMode {
			log.Printf("flrig: mode changed from %s to %s", oldMode, mode)
		}
		if f.modeCallback != nil {
			f.modeCallback(mode)
		} else {
			log.Printf("flrig: mode callback is nil!")
		}
	}
	if ptt != oldPTT || isFirstPoll {
		if ptt != oldPTT {
			log.Printf("flrig: PTT changed from %v to %v", oldPTT, ptt)
		}
		if f.pttCallback != nil {
			f.pttCallback(ptt)
		} else {
			log.Printf("flrig: PTT callback is nil!")
		}
	}
}

// GetCachedFrequency returns the cached frequency
func (f *FlrigClient) GetCachedFrequency() int {
	f.cacheMu.RLock()
	defer f.cacheMu.RUnlock()
	return f.cachedFrequency
}

// GetCachedMode returns the cached mode
func (f *FlrigClient) GetCachedMode() string {
	f.cacheMu.RLock()
	defer f.cacheMu.RUnlock()
	return f.cachedMode
}

// GetCachedPTT returns the cached PTT state
func (f *FlrigClient) GetCachedPTT() bool {
	f.cacheMu.RLock()
	defer f.cacheMu.RUnlock()
	return f.cachedPTT
}

// SetCallbacks sets the callback functions
func (f *FlrigClient) SetCallbacks(
	frequencyCallback func(int),
	modeCallback func(string),
	pttCallback func(bool),
	errorCallback func(string),
) {
	f.mu.Lock()
	defer f.mu.Unlock()

	f.frequencyCallback = frequencyCallback
	f.modeCallback = modeCallback
	f.pttCallback = pttCallback
	f.errorCallback = errorCallback
}

// callMethod makes an XML-RPC method call
func (f *FlrigClient) callMethod(method string, params []interface{}) (interface{}, error) {
	// Build XML-RPC request manually to ensure proper formatting
	var xmlRequest bytes.Buffer
	xmlRequest.WriteString(xml.Header)
	xmlRequest.WriteString("<methodCall>")
	xmlRequest.WriteString("<methodName>")
	xmlRequest.WriteString(method)
	xmlRequest.WriteString("</methodName>")

	if len(params) > 0 {
		xmlRequest.WriteString("<params>")
		for _, param := range params {
			xmlRequest.WriteString("<param><value>")
			switch v := param.(type) {
			case string:
				xmlRequest.WriteString("<string>")
				xml.EscapeText(&xmlRequest, []byte(v))
				xmlRequest.WriteString("</string>")
			case int:
				xmlRequest.WriteString(fmt.Sprintf("<int>%d</int>", v))
			case float64:
				xmlRequest.WriteString(fmt.Sprintf("<double>%f</double>", v))
			case bool:
				if v {
					xmlRequest.WriteString("<boolean>1</boolean>")
				} else {
					xmlRequest.WriteString("<boolean>0</boolean>")
				}
			default:
				xmlRequest.WriteString(fmt.Sprintf("<string>%v</string>", v))
			}
			xmlRequest.WriteString("</value></param>")
		}
		xmlRequest.WriteString("</params>")
	}

	xmlRequest.WriteString("</methodCall>")

	// Make HTTP POST request
	url := fmt.Sprintf("http://%s:%d", f.host, f.port)
	resp, err := f.client.Post(url, "text/xml", &xmlRequest)
	if err != nil {
		return nil, fmt.Errorf("HTTP request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("HTTP error: %s", resp.Status)
	}

	// Read response
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	// Parse the response manually to extract the value
	return parseXMLRPCResponse(body)
}

// parseXMLRPCResponse parses an XML-RPC response and extracts the value
func parseXMLRPCResponse(data []byte) (interface{}, error) {
	// Simple XML parsing to extract the value
	// Look for <value>...</value> in the response

	// Check for fault first
	if bytes.Contains(data, []byte("<fault>")) {
		return nil, fmt.Errorf("XML-RPC fault in response")
	}

	// Find the value content
	valueStart := bytes.Index(data, []byte("<value>"))
	if valueStart == -1 {
		return nil, fmt.Errorf("no value found in response")
	}
	valueStart += len("<value>")

	valueEnd := bytes.Index(data[valueStart:], []byte("</value>"))
	if valueEnd == -1 {
		return nil, fmt.Errorf("malformed value in response")
	}

	valueContent := data[valueStart : valueStart+valueEnd]

	// Check for different value types
	if bytes.Contains(valueContent, []byte("<string>")) {
		start := bytes.Index(valueContent, []byte("<string>")) + len("<string>")
		end := bytes.Index(valueContent, []byte("</string>"))
		if end > start {
			return string(valueContent[start:end]), nil
		}
	}

	if bytes.Contains(valueContent, []byte("<double>")) {
		start := bytes.Index(valueContent, []byte("<double>")) + len("<double>")
		end := bytes.Index(valueContent, []byte("</double>"))
		if end > start {
			return string(valueContent[start:end]), nil
		}
	}

	if bytes.Contains(valueContent, []byte("<int>")) || bytes.Contains(valueContent, []byte("<i4>")) {
		var start, end int
		if bytes.Contains(valueContent, []byte("<int>")) {
			start = bytes.Index(valueContent, []byte("<int>")) + len("<int>")
			end = bytes.Index(valueContent, []byte("</int>"))
		} else {
			start = bytes.Index(valueContent, []byte("<i4>")) + len("<i4>")
			end = bytes.Index(valueContent, []byte("</i4>"))
		}
		if end > start {
			return string(valueContent[start:end]), nil
		}
	}

	if bytes.Contains(valueContent, []byte("<boolean>")) {
		start := bytes.Index(valueContent, []byte("<boolean>")) + len("<boolean>")
		end := bytes.Index(valueContent, []byte("</boolean>"))
		if end > start {
			val := string(valueContent[start:end])
			return val == "1" || val == "true", nil
		}
	}

	// If no type tag, treat as string
	return string(bytes.TrimSpace(valueContent)), nil
}

// XML-RPC structures for marshaling/unmarshaling

type xmlrpcMethodCall struct {
	XMLName    xml.Name     `xml:"methodCall"`
	MethodName string       `xml:"methodName"`
	Params     xmlrpcParams `xml:"params"`
}

type xmlrpcParams struct {
	Params []xmlrpcParam `xml:"param"`
}

type xmlrpcParam struct {
	Value xmlrpcValue `xml:"value"`
}

type xmlrpcValue struct {
	Data interface{} `xml:",any"`
}

type xmlrpcMethodResponse struct {
	XMLName xml.Name     `xml:"methodResponse"`
	Params  xmlrpcParams `xml:"params"`
	Fault   xmlrpcFault  `xml:"fault"`
}

type xmlrpcFault struct {
	Value xmlrpcFaultValue `xml:"value"`
}

type xmlrpcFaultValue struct {
	Struct xmlrpcStruct `xml:"struct"`
}

type xmlrpcStruct struct {
	Members []xmlrpcMember `xml:"member"`
}

type xmlrpcMember struct {
	Name  string      `xml:"name"`
	Value xmlrpcValue `xml:"value"`
}

File: clients/python/flrig_control.py
#!/usr/bin/env python3
"""
flrig XML-RPC Client for Radio Control

A Python client for controlling radios via flrig's XML-RPC interface over TCP/IP.
Implements bidirectional frequency and mode synchronization with physical radios.

Usage:
    from flrig_control import FlrigClient
    
    # Connect to flrig
    rig = FlrigClient('localhost', 12345)
    rig.connect()
    
    # Get/Set frequency
    freq = rig.get_frequency()
    rig.set_frequency(14074000)
    
    # Get/Set mode
    mode = rig.get_mode()
    rig.set_mode('USB')
    
    # Disconnect
    rig.disconnect()
"""

import xmlrpc.client
import threading
import queue
import time
from typing import Optional, Callable, Any


class FlrigClient:
    """Simple flrig XML-RPC client for controlling external radios over TCP/IP.
    
    This class implements the flrig XML-RPC protocol for radio control.
    
    Protocol details:
    - Uses XML-RPC over HTTP
    - Default port is 12345
    - Communication happens over TCP/IP
    
    Attributes:
        host: Hostname or IP address of flrig server
        port: TCP port number (default: 12345)
        connected: Connection status
        vfo: Current VFO ('A' or 'B')
    """
    
    def __init__(self, host: str = '127.0.0.1', port: int = 12345, vfo: str = 'A'):
        """Initialize flrig client.
        
        Args:
            host: Hostname or IP address of flrig server
            port: TCP port number (default: 12345)
            vfo: VFO to use ('A' or 'B')
        """
        self.host = host
        self.port = port
        self.vfo = vfo.upper()
        self.server: Optional[xmlrpc.client.ServerProxy] = None
        self.connected = False
    
    def connect(self) -> bool:
        """Connect to flrig XML-RPC server.
        
        Returns:
            True if connection successful
            
        Raises:
            ConnectionError: If connection fails
        """
        try:
            self.server = xmlrpc.client.ServerProxy(
                f"http://{self.host}:{self.port}",
                allow_none=True
            )
            # Test connection by listing methods
            methods = self.server.system.listMethods()
            self.connected = True
            return True
        except Exception as e:
            self.connected = False
            raise ConnectionError(f"Failed to connect to flrig at {self.host}:{self.port}: {e}")
    
    def disconnect(self):
        """Disconnect from flrig server."""
        self.connected = False
        self.server = None
    
    def get_frequency(self) -> int:
        """Get current frequency in Hz.
        
        Returns:
            Frequency in Hz
            
        Raises:
            ConnectionError: If communication fails
            ValueError: If response cannot be parsed
        """
        if not self.connected or not self.server:
            raise ConnectionError("Not connected to flrig")
        
        try:
            freq_str = self.server.rig.get_vfo()
            return int(float(freq_str))
        except Exception as e:
            self.connected = False
            raise ConnectionError(f"Communication error: {e}")
    
    def set_frequency(self, freq_hz: int):
        """Set frequency in Hz.
        
        Args:
            freq_hz: Frequency in Hz
            
        Raises:
            ConnectionError: If communication fails
        """
        if not self.connected or not self.server:
            raise ConnectionError("Not connected to flrig")
        
        try:
            self.server.rig.set_vfo(float(freq_hz))
        except Exception as e:
            self.connected = False
            raise ConnectionError(f"Communication error: {e}")
    
    def get_mode(self) -> str:
        """Get current mode.
        
        Returns:
            Mode string (e.g., 'USB', 'LSB', 'CW', 'AM', 'FM')
            
        Raises:
            ConnectionError: If communication fails
        """
        if not self.connected or not self.server:
            raise ConnectionError("Not connected to flrig")
        
        try:
            mode = self.server.rig.get_mode()
            return mode if mode else 'Unknown'
        except Exception as e:
            self.connected = False
            raise ConnectionError(f"Communication error: {e}")
    
    def set_mode(self, mode: str):
        """Set mode (USB, LSB, CW, etc.).
        
        Args:
            mode: Mode string (e.g., 'USB', 'LSB', 'CW', 'AM', 'FM')
            
        Raises:
            ConnectionError: If communication fails
        """
        if not self.connected or not self.server:
            raise ConnectionError("Not connected to flrig")
        
        try:
            self.server.rig.set_mode(mode.upper())
        except Exception as e:
            self.connected = False
            raise ConnectionError(f"Communication error: {e}")
    
    def get_vfo(self) -> str:
        """Get current VFO.
        
        Returns:
            VFO string ('A' or 'B')
        """
        return self.vfo
    
    def set_vfo(self, vfo: str):
        """Set VFO and switch to it in flrig.
        
        Args:
            vfo: VFO string ('A' or 'B')
            
        Raises:
            ConnectionError: If communication fails
        """
        if not self.connected or not self.server:
            raise ConnectionError("Not connected to flrig")
        
        vfo = vfo.upper()
        try:
            # Use rig.set_AB to switch VFO in flrig
            # 'A' = VFO A, 'B' = VFO B
            self.server.rig.set_AB(vfo)
            self.vfo = vfo
        except Exception as e:
            self.connected = False
            raise ConnectionError(f"Communication error: {e}")
    
    def get_ptt(self) -> bool:
        """Get PTT (Push-To-Talk) status.
        
        Returns:
            True if PTT is active (transmitting), False otherwise
            
        Raises:
            ConnectionError: If communication fails
        """
        if not self.connected or not self.server:
            raise ConnectionError("Not connected to flrig")
        
        try:
            ptt = self.server.rig.get_ptt()
            return bool(ptt)
        except Exception as e:
            self.connected = False
            raise ConnectionError(f"Communication error: {e}")
    
    def set_ptt(self, state: bool):
        """Set PTT (Push-To-Talk) state.
        
        Args:
            state: True to transmit, False to receive
            
        Raises:
            ConnectionError: If communication fails
        """
        if not self.connected or not self.server:
            raise ConnectionError("Not connected to flrig")
        
        try:
            self.server.rig.set_ptt(1 if state else 0)
        except Exception as e:
            self.connected = False
            raise ConnectionError(f"Communication error: {e}")
    
    def get_xcvr(self) -> str:
        """Get transceiver name.
        
        Returns:
            Transceiver name string
            
        Raises:
            ConnectionError: If communication fails
        """
        if not self.connected or not self.server:
            raise ConnectionError("Not connected to flrig")
        
        try:
            return self.server.rig.get_xcvr()
        except Exception as e:
            self.connected = False
            raise ConnectionError(f"Communication error: {e}")
    
    def get_modes(self) -> list:
        """Get available modes.
        
        Returns:
            List of available mode strings
            
        Raises:
            ConnectionError: If communication fails
        """
        if not self.connected or not self.server:
            raise ConnectionError("Not connected to flrig")
        
        try:
            return self.server.rig.get_modes()
        except Exception as e:
            self.connected = False
            raise ConnectionError(f"Communication error: {e}")
    
    def get_bw(self) -> tuple:
        """Get current bandwidth.
        
        Returns:
            Tuple of (bw1, bw2) bandwidth strings
            
        Raises:
            ConnectionError: If communication fails
        """
        if not self.connected or not self.server:
            raise ConnectionError("Not connected to flrig")
        
        try:
            result = self.server.rig.get_bw()
            if isinstance(result, list) and len(result) >= 2:
                return (result[1], result[0])  # (bw1, bw2)
            return ('', '')
        except Exception as e:
            self.connected = False
            raise ConnectionError(f"Communication error: {e}")
    
    def set_bw(self, bw2: int, bw1: int = -1):
        """Set bandwidth.
        
        Args:
            bw2: Bandwidth index for BW2
            bw1: Bandwidth index for BW1 (optional)
            
        Raises:
            ConnectionError: If communication fails
        """
        if not self.connected or not self.server:
            raise ConnectionError("Not connected to flrig")
        
        try:
            ival = bw2
            if bw1 > -1:
                ival = 256 * (bw1 + 128) + bw2
            self.server.rig.set_bw(ival)
        except Exception as e:
            self.connected = False
            raise ConnectionError(f"Communication error: {e}")


class ThreadedFlrigClient:
    """Thread-safe flrig client that runs operations in a background thread.
    
    This class wraps FlrigClient and moves all XML-RPC I/O to a background thread,
    preventing GUI blocking. Commands are queued and executed asynchronously, with
    callbacks for results.
    
    Attributes:
        host: Hostname or IP address of flrig server
        port: TCP port number (default: 12345)
        vfo: Current VFO ('A' or 'B')
        connected: Connection status
        running: Thread running status
    """
    
    def __init__(self, host: str = '127.0.0.1', port: int = 12345, vfo: str = 'A'):
        """Initialize threaded flrig client.
        
        Args:
            host: Hostname or IP address of flrig server
            port: TCP port number (default: 12345)
            vfo: VFO to use ('A' or 'B')
        """
        self.host = host
        self.port = port
        self.vfo = vfo.upper()
        self.rig = FlrigClient(host, port, vfo)
        self.connected = False
        self.running = False
        
        # Command queue for thread-safe operations
        self.command_queue: queue.Queue = queue.Queue()
        
        # Result callbacks
        self.frequency_callback: Optional[Callable[[int], None]] = None
        self.mode_callback: Optional[Callable[[str], None]] = None
        self.ptt_callback: Optional[Callable[[bool], None]] = None
        self.error_callback: Optional[Callable[[str], None]] = None
        
        # Worker thread
        self.worker_thread: Optional[threading.Thread] = None
        
        # Cached values for quick access (updated by polling)
        self._cached_frequency: Optional[int] = None
        self._cached_mode: Optional[str] = None
        self._cached_ptt: bool = False
        self._cache_lock = threading.Lock()
    
    def connect(self) -> bool:
        """Connect to flrig server and start worker thread.
        
        Returns:
            True if connection successful
            
        Raises:
            ConnectionError: If connection fails
        """
        # Connect the underlying client
        self.rig.connect()
        self.connected = True
        
        # Start worker thread
        self.running = True
        self.worker_thread = threading.Thread(target=self._worker_loop, daemon=True)
        self.worker_thread.start()
        
        return True
    
    def disconnect(self):
        """Disconnect from flrig server and stop worker thread."""
        self.running = False
        
        # Wait for worker thread to finish
        if self.worker_thread and self.worker_thread.is_alive():
            self.worker_thread.join(timeout=1.0)
        
        # Disconnect underlying client
        if self.connected:
            self.rig.disconnect()
            self.connected = False
    
    def _worker_loop(self):
        """Worker thread loop that processes commands from queue."""
        while self.running:
            try:
                # Get command from queue with timeout
                try:
                    cmd_type, args = self.command_queue.get(timeout=0.02)  # 20ms timeout
                except queue.Empty:
                    continue
                
                # Execute command
                try:
                    if cmd_type == 'get_frequency':
                        freq = self.rig.get_frequency()
                        with self._cache_lock:
                            self._cached_frequency = freq
                        if self.frequency_callback:
                            self.frequency_callback(freq)
                    
                    elif cmd_type == 'set_frequency':
                        freq_hz = args[0]
                        self.rig.set_frequency(freq_hz)
                        with self._cache_lock:
                            self._cached_frequency = freq_hz
                    
                    elif cmd_type == 'get_mode':
                        mode = self.rig.get_mode()
                        with self._cache_lock:
                            self._cached_mode = mode
                        if self.mode_callback:
                            self.mode_callback(mode)
                    
                    elif cmd_type == 'set_mode':
                        mode = args[0]
                        self.rig.set_mode(mode)
                        with self._cache_lock:
                            self._cached_mode = mode
                    
                    elif cmd_type == 'get_ptt':
                        ptt = self.rig.get_ptt()
                        with self._cache_lock:
                            self._cached_ptt = ptt
                        if self.ptt_callback:
                            self.ptt_callback(ptt)
                    
                    elif cmd_type == 'set_ptt':
                        state = args[0]
                        self.rig.set_ptt(state)
                        with self._cache_lock:
                            self._cached_ptt = state
                    
                    elif cmd_type == 'set_vfo':
                        vfo = args[0]
                        self.rig.set_vfo(vfo)
                        self.vfo = vfo
                    
                    elif cmd_type == 'poll':
                        # Poll all values
                        freq = self.rig.get_frequency()
                        mode = self.rig.get_mode()
                        ptt = self.rig.get_ptt()
                        
                        with self._cache_lock:
                            self._cached_frequency = freq
                            self._cached_mode = mode
                            self._cached_ptt = ptt
                        
                        # Trigger callbacks
                        if self.frequency_callback:
                            self.frequency_callback(freq)
                        if self.mode_callback:
                            self.mode_callback(mode)
                        if self.ptt_callback:
                            self.ptt_callback(ptt)
                
                except Exception as e:
                    if self.error_callback:
                        self.error_callback(str(e))
                
                finally:
                    self.command_queue.task_done()
            
            except Exception:
                # Catch any unexpected errors to keep thread alive
                pass
    
    def get_frequency(self) -> Optional[int]:
        """Get cached frequency (non-blocking).
        
        Returns:
            Cached frequency in Hz, or None if not yet polled
        """
        with self._cache_lock:
            return self._cached_frequency
    
    def set_frequency(self, freq_hz: int):
        """Queue frequency change command (non-blocking).
        
        Args:
            freq_hz: Frequency in Hz
        """
        self.command_queue.put(('set_frequency', (freq_hz,)))
    
    def get_mode(self) -> Optional[str]:
        """Get cached mode (non-blocking).
        
        Returns:
            Cached mode string, or None if not yet polled
        """
        with self._cache_lock:
            return self._cached_mode
    
    def set_mode(self, mode: str):
        """Queue mode change command (non-blocking).
        
        Args:
            mode: Mode string (e.g., 'USB', 'LSB', 'CW', 'AM', 'FM')
        """
        self.command_queue.put(('set_mode', (mode,)))
    
    def get_vfo(self) -> str:
        """Get current VFO (non-blocking).
        
        Returns:
            Current VFO ('A' or 'B')
        """
        return self.vfo
    
    def set_vfo(self, vfo: str):
        """Queue VFO change command (non-blocking).
        
        Args:
            vfo: VFO string ('A' or 'B')
        """
        self.command_queue.put(('set_vfo', (vfo,)))
    
    def get_ptt(self) -> bool:
        """Get cached PTT status (non-blocking).
        
        Returns:
            Cached PTT state
        """
        with self._cache_lock:
            return self._cached_ptt
    
    def set_ptt(self, state: bool):
        """Queue PTT change command (non-blocking).
        
        Args:
            state: True to transmit, False to receive
        """
        self.command_queue.put(('set_ptt', (state,)))
    
    def poll(self):
        """Queue a poll command to update all cached values (non-blocking)."""
        self.command_queue.put(('poll', ()))
    
    def set_callbacks(self, 
                     frequency_callback: Optional[Callable[[int], None]] = None,
                     mode_callback: Optional[Callable[[str], None]] = None,
                     ptt_callback: Optional[Callable[[bool], None]] = None,
                     error_callback: Optional[Callable[[str], None]] = None):
        """Set callbacks for value changes.
        
        Args:
            frequency_callback: Called when frequency is polled
            mode_callback: Called when mode is polled
            ptt_callback: Called when PTT status is polled
            error_callback: Called when an error occurs
        """
        self.frequency_callback = frequency_callback
        self.mode_callback = mode_callback
        self.ptt_callback = ptt_callback
        self.error_callback = error_callback


def main():
    """Example usage of FlrigClient."""
    import sys
    
    # Parse command line arguments
    host = sys.argv[1] if len(sys.argv) > 1 else '127.0.0.1'
    port = int(sys.argv[2]) if len(sys.argv) > 2 else 12345
    
    try:
        # Connect to flrig
        print(f"Connecting to flrig at {host}:{port}...")
        rig = FlrigClient(host, port)
        rig.connect()
        print("✓ Connected")
        
        # Get transceiver name
        try:
            xcvr = rig.get_xcvr()
            print(f"Transceiver: {xcvr}")
        except:
            print("Could not get transceiver name")
        
        # Get current frequency
        freq = rig.get_frequency()
        print(f"Current frequency: {freq/1e6:.6f} MHz")
        
        # Get current mode
        mode = rig.get_mode()
        print(f"Current mode: {mode}")
        
        # Example: Set frequency to 14.074 MHz
        print("\nSetting frequency to 14.074 MHz...")
        rig.set_frequency(14074000)
        
        # Verify
        time.sleep(0.5)
        freq = rig.get_frequency()
        print(f"New frequency: {freq/1e6:.6f} MHz")
        
        # Example: Set mode to USB
        print("\nSetting mode to USB...")
        rig.set_mode('USB')
        
        # Verify
        time.sleep(0.5)
        mode = rig.get_mode()
        print(f"New mode: {mode}")
        
        # Disconnect
        rig.disconnect()
        print("\n✓ Disconnected")
        
    except ConnectionError as e:
        print(f"ERROR: {e}", file=sys.stderr)
        sys.exit(1)
    except KeyboardInterrupt:
        print("\nInterrupted by user")
        if rig.connected:
            rig.disconnect()
        sys.exit(0)


if __name__ == '__main__':
    main()